feat: 배송지 기능 구현

MCshin00 opened the pull request:
### PR 타입
- [x] 기능 추가
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트

### 반영 브랜치
- feature/address-api -> main

### 변경 사항
- 배송지 관련 CRUD 기능 구현 
- 기본 배송지 지정 기능 구현

MCshin00 committed:
feat: 배송지 기능 구현

배송지 등록, 수정, 목록 조회 기능 구현

MCshin00 committed:
refactor: 배송지 수정 API 예외처리 추가

본인의 배송지가 아닌 addressId로 수정을 시도하면 예외가 발생하도록 변경

MCshin00 committed:
feat: 현재(기본) 배송지 지정 기능 구현

MCshin00 committed:
refactor: AddressService 예외처리 추가 및 파라미터 변경

service의 updateAddress 메서드에서 해당 주소가 없는 경우와 해당 주소가 자신의 것이 아닌 경우로 나누어 예외처리 하도록 변경
username을 파라미터로 받고 User를 조회하던 로직을 제거하고 AuthenticationPrincipal를 통해 User를 직접 파라미터로 받는 방식으로 변경

MCshin00 committed:
feat: 배송지 삭제 기능 구현

MCshin00 committed:
refactor: 사용하지 않는 함수 제거 및 필요한 멤버 추가

AddressRepository에서 사용하지 않는 findByIdAndUser 메서드 제거
setCurrentAddress 메서드에서 save 메서드 호출을 위해 UserRepository 선언을 다시 추가

MCshin00 committed:
refactor: 에러 메세지를 Enum으로 관리하도록 변경

MCshin00 committed:
refactor: 응답 body를 addressIdResponseDto로 넘기도록 변경

addressIdResponseDto를 통해 생성, 수정, 삭제 처리한 주소의 Id를 body에 담아 전달하도록 변경

strongcookdas commented on blackWhiteDeliveryDriver/src/main/java/com/sparta/blackwhitedeliverydriver/controller/AddressController.java:
```diff
@@ -0,0 +1,79 @@
+@Slf4j
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/users/address")
```
base uri 적용이 안된 것 같습니다. `/api/v1`

MCshin00 replied:
적용했습니다!

strongcookdas commented on blackWhiteDeliveryDriver/src/main/java/com/sparta/blackwhitedeliverydriver/entity/Address.java:
```diff
@@ -0,0 +1,58 @@
+public class Address extends BaseEntity{
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
```
ch1 요구사항에 모든 주요 엔티티의 식별자는 UUID를 사용하라고 나와 있어서 확인해보시면 좋을 것 같아요.
데이터베이스 설계 파트 부분에 기록되어 있습니다.
[ch1 프로젝트 노션 링크](https://teamsparta.notion.site/Chapter-1-AI-1352dc3ef5148040ab1ed75e8e95e72e)

MCshin00 replied:
깜빡했습니다. UUID로 변경하겠습니다.

strongcookdas commented on blackWhiteDeliveryDriver/src/main/java/com/sparta/blackwhitedeliverydriver/service/AddressService.java:
```diff
@@ -0,0 +1,97 @@
+            throw new AccessDeniedException(ExceptionMessage.NOT_ALLOWED_API.getMessage());
+        }
+
+        address.setZipNum(requestDto.getZipNum());
+        address.setCity(requestDto.getCity());
+        address.setDistrict(requestDto.getDistrict());
+        address.setStreetName(requestDto.getStreetName());
+        address.setStreetNum(requestDto.getStreetNum());
+        address.setDetailAddr(requestDto.getDetailAddr());
```
address 엔티티에 setter를 이용해서 dto값을 대입하셨는데
정적 팩토리 메서드 중 `from` 네이밍을 사용하셔서 값을 대입하시면 더 좋을 것 같아 제안드립니다!
[정적 팩토리 메서드 참고 블로그](https://tecoble.techcourse.co.kr/post/2020-05-26-static-factory-method/)

그리고 entity에서는 `@setter` 어노테이션을 지양한다는 의견을 많이 들었어요. 그 이유가 누구든지 수정 접근이 가능하기 때문이라고 많이 그러는 것 같아요. 이것에 대해서도 한번 찾아보시면 좋을 것 같습니다!
[참고 블로그 링크](https://velog.io/@langoustine/setter-%EC%A7%80%EC%96%91-%EC%9D%B4%EC%9C%A0)

MCshin00 replied:
둘 다 고민하고 있던 부분이었는데, 지적 감사합니다. 우선 Address에 정적 팩토리 메서드 추가와 ``@setter`` 제거를 진행하고, User에서도 적용해보겠습니다.

strongcookdas reviewed (commented):
코드 리뷰 올립니다!

MCshin00 committed:
refactor: Address의 Id 타입을 UUID로 변경 및 base uri 변경

base uri api/v1/ 추가
@DeleteMapping("/{addressId}") 아래에 선언된 @PutMapping("/{addressId}") 제거
dto 클래스에 선언한 id를 UUID 타입으로 변경

MCshin00 committed:
fix: User를 비교해서 자신의 주소인지 판단하는 조건문의 버그 수정

address.getUser()는 User 엔티티를 반환하고, userDetails.getUser()는 UserDetailsImpl 객체를 반환함
따라서 equals() 메서드의 결과는 항상 false가 되어 검증을 제대로 수행하지 못하는 문제 발생
getUsername으로 비교하도록 수정

MCshin00 committed:
refactor: 정적 팩토리 메서드 추가 및 세터 제거

dto를 통해 Address Entity를 생성하는 정적 팩토리 메서드를 추가
update에서 필드 값 업데이트를 위해 사용하던 setter를 제거하고 update 메서드를 추가해 사용하도록 변경

MCshin00 committed:
refactor: ResponseDto에 정적 팩토리 메서드 추가 및 삭제 예외처리 추가

ResponseDto에 정적 팩토리 메서드를 추가하고 파라미터로 받은 Address 객체를 통해 Builder로 dto 객체 생성하게 변경
Repository의 함수에 Query를 직접 작성하고 deleted_by나 deleted_date 필드가 null이 아닌 값만 가져오도록 변경
Service에 Delete 여부를 체크하는 함수를 추가하고 예외를 throw 하도록 변경

MCshin00 committed:
refactor: deleteAddress에 삭제여부 검증 추가

MCshin00 committed:
refactor: 예외처리 메세지 enum으로 처리하도록 변경
